Application using Hibernate offline startup tries to connect to database in DEV mode on startup

### Describe the bug

I am running Quarkus application with Hibernate ORM and use offline startup, however when I start DEV mode and database is not ready, I can see validation exceptions due to this https://github.com/quarkusio/quarkus/blob/28fed6bfce97c09d6c573feb87d33b8959bda073/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/HibernateOrmRecorder.java#L272. Applications still starts, but the application log is not pretty and users may think they have done something wrong.

I'd also like to point out, that since my scenario is that the database credentials are provided by incoming HTTP request, attempt to connect to database on startup could hardly pass even if the database was running (hence the wrong credentials in my case).

### Expected behavior

For the offline startup, I'd expect that I can start app without database and start database later.

### Actual behavior

Application starts and I can see following messages in my log:

```
19:39:48,737 INFO  [app] 19:39:47,207 Datasource 'app_scope_credentials': Connection to localhost:5432 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.
19:39:48,737 INFO  [app] 19:39:47,208 HHH000247: ErrorCode: 0, SQLState: 08001
19:39:48,737 INFO  [app] 19:39:47,208 HHH000247: ErrorCode: 0, SQLState: 08001
19:39:48,737 INFO  [app] 19:39:47,208 Connection to localhost:5432 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.
19:39:48,737 INFO  [app] 19:39:47,208 Connection to localhost:5432 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.
19:39:48,737 INFO  [app] 19:39:47,209 Failed to run post-boot validation: org.hibernate.exception.JDBCConnectionException: Unable to open JDBC Connection for DDL execution [Connection to localhost:5432 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.] [n/a]
19:39:48,737 INFO  [app] 	at org.hibernate.exception.internal.SQLStateConversionDelegate.convert(SQLStateConversionDelegate.java:80)
19:39:48,737 INFO  [app] 	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:34)
19:39:48,737 INFO  [app] 	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:115)
19:39:48,738 INFO  [app] 	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:101)
19:39:48,738 INFO  [app] 	at org.hibernate.resource.transaction.backend.jta.internal.DdlTransactionIsolatorJtaImpl.getIsolatedConnection(DdlTransactionIsolatorJtaImpl.java:76)
19:39:48,738 INFO  [app] 	at org.hibernate.resource.transaction.backend.jta.internal.DdlTransactionIsolatorJtaImpl.getIsolatedConnection(DdlTransactionIsolatorJtaImpl.java:65)
19:39:48,738 INFO  [app] 	at org.hibernate.tool.schema.internal.exec.ImprovedExtractionContextImpl.getJdbcConnection(ImprovedExtractionContextImpl.java:61)
19:39:48,738 INFO  [app] 	at org.hibernate.tool.schema.extract.spi.ExtractionContext.getQueryResults(ExtractionContext.java:40)
19:39:48,738 INFO  [app] 	at org.hibernate.tool.schema.extract.internal.SequenceInformationExtractorLegacyImpl.extractMetadata(SequenceInformationExtractorLegacyImpl.java:36)
19:39:48,738 INFO  [app] 	at org.hibernate.tool.schema.extract.internal.DatabaseInformationImpl.initializeSequences(DatabaseInformationImpl.java:64)
19:39:48,738 INFO  [app] 	at org.hibernate.tool.schema.extract.internal.DatabaseInformationImpl.<init>(DatabaseInformationImpl.java:58)
19:39:48,738 INFO  [app] 	at org.hibernate.tool.schema.internal.Helper.buildDatabaseInformation(Helper.java:178)
19:39:48,738 INFO  [app] 	at org.hibernate.tool.schema.internal.AbstractSchemaValidator.doValidation(AbstractSchemaValidator.java:67)
19:39:48,738 INFO  [app] 	at io.quarkus.hibernate.orm.runtime.schema.SchemaManagementIntegrator.runPostBootValidation(SchemaManagementIntegrator.java:140)
19:39:48,738 INFO  [app] 	at io.quarkus.hibernate.orm.runtime.HibernateOrmRecorder$13.run(HibernateOrmRecorder.java:285)
19:39:48,738 INFO  [app] 	at java.base/java.lang.Thread.run(Thread.java:1583)
19:39:48,739 INFO  [app] Caused by: org.postgresql.util.PSQLException: Connection to localhost:5432 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.
19:39:48,739 INFO  [app] 	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:373)
19:39:48,739 INFO  [app] 	at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:57)
19:39:48,739 INFO  [app] 	at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:279)
19:39:48,739 INFO  [app] 	at org.postgresql.Driver.makeConnection(Driver.java:448)
19:39:48,739 INFO  [app] 	at org.postgresql.Driver.connect(Driver.java:298)
19:39:48,739 INFO  [app] 	at io.agroal.pool.ConnectionFactory.createConnection(ConnectionFactory.java:225)
19:39:48,739 INFO  [app] 	at io.agroal.pool.Poolless.createConnection(Poolless.java:400)
19:39:48,739 INFO  [app] 	at io.agroal.pool.Poolless.handlerFromSharedCache(Poolless.java:264)
19:39:48,739 INFO  [app] 	at io.agroal.pool.Poolless.getConnection(Poolless.java:238)
19:39:48,739 INFO  [app] 	at io.agroal.pool.DataSource.getConnection(DataSource.java:86)
19:39:48,739 INFO  [app] 	at io.agroal.api.AgroalDataSource_HM0rs3xICl1FGiizjDnTbiZ4qCQ_Synthetic_ClientProxy.getConnection(Unknown Source)
19:39:48,739 INFO  [app] 	at io.quarkus.hibernate.orm.runtime.customized.QuarkusConnectionProvider.getConnection(QuarkusConnectionProvider.java:27)
19:39:48,740 INFO  [app] 	at io.quarkus.hibernate.orm.runtime.tenant.DataSourceTenantConnectionResolver$SchemaTenantConnectionProvider.getConnection(DataSourceTenantConnectionResolver.java:89)
19:39:48,740 INFO  [app] 	at org.hibernate.engine.jdbc.connections.spi.AbstractMultiTenantConnectionProvider.getAnyConnection(AbstractMultiTenantConnectionProvider.java:27)
19:39:48,740 INFO  [app] 	at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator$MultiTenantConnectionProviderJdbcConnectionAccess.obtainConnection(JdbcEnvironmentInitiator.java:526)
19:39:48,740 INFO  [app] 	at org.hibernate.resource.transaction.backend.jta.internal.DdlTransactionIsolatorJtaImpl.getIsolatedConnection(DdlTransactionIsolatorJtaImpl.java:72)
19:39:48,740 INFO  [app] 	... 11 more
19:39:48,740 INFO  [app] Caused by: java.net.ConnectException: Connection refused
19:39:48,740 INFO  [app] 	at java.base/sun.nio.ch.Net.pollConnect(Native Method)
19:39:48,740 INFO  [app] 	at java.base/sun.nio.ch.Net.pollConnectNow(Net.java:682)
19:39:48,740 INFO  [app] 	at java.base/sun.nio.ch.NioSocketImpl.timedFinishConnect(NioSocketImpl.java:542)
19:39:48,740 INFO  [app] 	at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:592)
19:39:48,740 INFO  [app] 	at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:327)
19:39:48,740 INFO  [app] 	at java.base/java.net.Socket.connect(Socket.java:751)
19:39:48,740 INFO  [app] 	at org.postgresql.core.PGStream.createSocket(PGStream.java:261)
19:39:48,740 INFO  [app] 	at org.postgresql.core.PGStream.<init>(PGStream.java:122)
19:39:48,740 INFO  [app] 	at org.postgresql.core.v3.ConnectionFactoryImpl.tryConnect(ConnectionFactoryImpl.java:146)
19:39:48,741 INFO  [app] 	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:289)
19:39:48,741 INFO  [app] 	... 26 more
19:39:48,741 INFO  [app] 19:39:47,209 Failed to run post-boot validation: org.hibernate.exception.JDBCConnectionException: Unable to open JDBC Connection for DDL execution [Connection to localhost:5432 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.] [n/a]
19:39:48,741 INFO  [app] 	at org.hibernate.exception.internal.SQLStateConversionDelegate.convert(SQLStateConversionDelegate.java:80)
19:39:48,741 INFO  [app] 	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:34)
19:39:48,741 INFO  [app] 	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:115)
19:39:48,741 INFO  [app] 	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:101)
19:39:48,741 INFO  [app] 	at org.hibernate.resource.transaction.backend.jta.internal.DdlTransactionIsolatorJtaImpl.getIsolatedConnection(DdlTransactionIsolatorJtaImpl.java:76)
19:39:48,741 INFO  [app] 	at org.hibernate.resource.transaction.backend.jta.internal.DdlTransactionIsolatorJtaImpl.getIsolatedConnection(DdlTransactionIsolatorJtaImpl.java:65)
19:39:48,741 INFO  [app] 	at org.hibernate.tool.schema.internal.exec.ImprovedExtractionContextImpl.getJdbcConnection(ImprovedExtractionContextImpl.java:61)
19:39:48,741 INFO  [app] 	at org.hibernate.tool.schema.extract.spi.ExtractionContext.getQueryResults(ExtractionContext.java:40)
19:39:48,741 INFO  [app] 	at org.hibernate.tool.schema.extract.internal.SequenceInformationExtractorLegacyImpl.extractMetadata(SequenceInformationExtractorLegacyImpl.java:36)
19:39:48,742 INFO  [app] 	at org.hibernate.tool.schema.extract.internal.DatabaseInformationImpl.initializeSequences(DatabaseInformationImpl.java:64)
19:39:48,742 INFO  [app] 	at org.hibernate.tool.schema.extract.internal.DatabaseInformationImpl.<init>(DatabaseInformationImpl.java:58)
19:39:48,742 INFO  [app] 	at org.hibernate.tool.schema.internal.Helper.buildDatabaseInformation(Helper.java:178)
19:39:48,742 INFO  [app] 	at org.hibernate.tool.schema.internal.AbstractSchemaValidator.doValidation(AbstractSchemaValidator.java:67)
19:39:48,742 INFO  [app] 	at io.quarkus.hibernate.orm.runtime.schema.SchemaManagementIntegrator.runPostBootValidation(SchemaManagementIntegrator.java:140)
19:39:48,742 INFO  [app] 	at io.quarkus.hibernate.orm.runtime.HibernateOrmRecorder$13.run(HibernateOrmRecorder.java:285)
19:39:48,742 INFO  [app] 	at java.base/java.lang.Thread.run(Thread.java:1583)
19:39:48,742 INFO  [app] Caused by: org.postgresql.util.PSQLException: Connection to localhost:5432 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.
19:39:48,742 INFO  [app] 	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:373)
19:39:48,742 INFO  [app] 	at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:57)
19:39:48,742 INFO  [app] 	at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:279)
19:39:48,742 INFO  [app] 	at org.postgresql.Driver.makeConnection(Driver.java:448)
19:39:48,742 INFO  [app] 	at org.postgresql.Driver.connect(Driver.java:298)
19:39:48,742 INFO  [app] 	at io.agroal.pool.ConnectionFactory.createConnection(ConnectionFactory.java:225)
19:39:48,742 INFO  [app] 	at io.agroal.pool.Poolless.createConnection(Poolless.java:400)
19:39:48,742 INFO  [app] 	at io.agroal.pool.Poolless.handlerFromSharedCache(Poolless.java:264)
19:39:48,743 INFO  [app] 	at io.agroal.pool.Poolless.getConnection(Poolless.java:238)
19:39:48,743 INFO  [app] 	at io.agroal.pool.DataSource.getConnection(DataSource.java:86)
19:39:48,743 INFO  [app] 	at io.quarkus.ts.hibernate.startup.offline.pu.ownconnection.orm.CustomTenantConnectionResolver$CustomConnectionProvider.getConnection(CustomTenantConnectionResolver.java:91)
19:39:48,743 INFO  [app] 	at org.hibernate.engine.jdbc.connections.spi.AbstractMultiTenantConnectionProvider.getAnyConnection(AbstractMultiTenantConnectionProvider.java:27)
19:39:48,743 INFO  [app] 	at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator$MultiTenantConnectionProviderJdbcConnectionAccess.obtainConnection(JdbcEnvironmentInitiator.java:526)
19:39:48,743 INFO  [app] 	at org.hibernate.resource.transaction.backend.jta.internal.DdlTransactionIsolatorJtaImpl.getIsolatedConnection(DdlTransactionIsolatorJtaImpl.java:72)
19:39:48,743 INFO  [app] 	... 11 more
19:39:48,743 INFO  [app] Caused by: java.net.ConnectException: Connection refused
19:39:48,743 INFO  [app] 	at java.base/sun.nio.ch.Net.pollConnect(Native Method)
19:39:48,743 INFO  [app] 	at java.base/sun.nio.ch.Net.pollConnectNow(Net.java:682)
19:39:48,743 INFO  [app] 	at java.base/sun.nio.ch.NioSocketImpl.timedFinishConnect(NioSocketImpl.java:542)
19:39:48,743 INFO  [app] 	at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:592)
19:39:48,743 INFO  [app] 	at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:327)
19:39:48,743 INFO  [app] 	at java.base/java.net.Socket.connect(Socket.java:751)
19:39:48,743 INFO  [app] 	at org.postgresql.core.PGStream.createSocket(PGStream.java:261)
19:39:48,743 INFO  [app] 	at org.postgresql.core.PGStream.<init>(PGStream.java:122)
19:39:48,743 INFO  [app] 	at org.postgresql.core.v3.ConnectionFactoryImpl.tryConnect(ConnectionFactoryImpl.java:146)
19:39:48,743 INFO  [app] 	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:289)
19:39:48,744 INFO  [app] 	... 24 more
```

### How to Reproduce?

Reproducer:

```
git clone git@github.com:michalvavrik/quarkus-test-suite.git -b reproducer/dev-mode-hibernate-offline-startup
cd quarkus-test-suite/hibernate/hibernate-offline-startup
mvn clean verify -Dit.test=DevModeHibernateOfflineStartupIT
```

For standalone Quarkus app without PostgreSQL, use:

```
git clone git@github.com:michalvavrik/hibernate-offline-startup-dev-mode-reproducer.git
cd hibernate-offline-startup-dev-mode-reproducer/
quarkus dev
```

### Output of `uname -a` or `ver`

Fedora 41

### Output of `java -version`

Temurin 21

### Quarkus version or git rev

999-SNAPSHOT

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Maven 3.9.9

### Additional information

_No response_

--- a/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/offline/StartOfflineDevModeTest.java
+++ b/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/offline/StartOfflineDevModeTest.java
@@ -1,0 +1,54 @@
+package io.quarkus.hibernate.orm.offline;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.hamcrest.Matchers.is;
+
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.Produces;
+import jakarta.ws.rs.core.MediaType;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.hibernate.orm.MyEntity;
+import io.quarkus.test.QuarkusDevModeTest;
+import io.restassured.RestAssured;
+
+/**
+ * Test that an application is running
+ * even if the database is offline when the application starts
+ * in DEV mode
+ */
+public class StartOfflineDevModeTest {
+
+    @RegisterExtension
+    static QuarkusDevModeTest runner = new QuarkusDevModeTest()
+            .withApplicationRoot((jar) -> jar
+                    .addClass(MyEntity.class)
+                    .addClass(GreetingResource.class)
+                    .addAsResource("application-start-offline.properties", "application.properties"))
+            .setLogRecordPredicate(record -> true);
+
+    @Test
+    public void testUnitSchemaManagementStrategyIsNone() {
+        RestAssured.when().get("/hello").then()
+                .statusCode(200)
+                .body(is("DB is offline but application is running"));
+
+        assertThat(runner.getLogRecords())
+                .map(l -> l.getMessage())
+                .doesNotContain("Failed to run post-boot validation");
+    }
+
+    @Path("/hello")
+    public static class GreetingResource {
+
+        @GET
+        @Produces(MediaType.TEXT_PLAIN)
+        public String hello() {
+            return "DB is offline but application is running";
+        }
+    }
+
+}

--- a/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/HibernateOrmRecorder.java
+++ b/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/HibernateOrmRecorder.java
@@ -290,8 +290,10 @@ public class HibernateOrmRecorder {
         String schemaManagementStrategy = hibernateOrmRuntimeConfigPersistenceUnit.database().generation().generation()
                 .orElse(hibernateOrmRuntimeConfigPersistenceUnit.schemaManagement().strategy());
 
-        //if hibernate is already managing the schema we don't do this
-        if (!"none".equals(schemaManagementStrategy)) {
+        boolean startsOffline = hibernateOrmRuntimeConfigPersistenceUnit.database().startOffline();
+
+        //if hibernate is already managing the schema or if we're in offline mode we don't do this
+        if (!"none".equals(schemaManagementStrategy) || startsOffline) {
             return;
         }
         new Thread(new Runnable() {

